chore(docs): organize installation doc (#141)
armory/docs: content/en/docs/installation/operator.md

---
title: Spinnaker Operator
weight: 1
---

Spinnaker Operator is a Kubernetes operator for Spinnaker that makes it easy to install, deploy, and upgrade any version of Spinnaker.

- Manage Spinnaker with `kubectl` like other applications.
- Expose Spinnaker via `LoadBalancer` or `Ingress` (optional)
- Keep secrets separate from your config. Store your config in `git` and have an easy Gitops workflow.
- Validate your configuration before applying it (with webhook validation).
- Store Spinnaker secrets in [Kubernetes secrets](https://github.com/armory/spinnaker-operator/blob/master/doc/managing-spinnaker.md#secrets-in-kubernetes-secrets).
- Gain total control over Spinnaker manifests with [`kustomize` style patching](https://github.com/armory/spinnaker-operator/blob/master/doc/options.md#speckustomize)
- Define Kubernetes accounts in `SpinnakerAccount` objects and store kubeconfig inline, in Kubernetes secrets, in s3, or GCS **(Experimental)**.

> - Deploy Spinnaker in an Istio controlled cluster **(Experimental)**

Reviewer: Would this list be better served in the product compatibility matrix? i.e. could we just link to our intro blog post here instead of listing all this out?

- You have `ValidatingAdmissionWebhook` enabled in the kube-apiserver. Alternatively, you can pass the `--disable-admission-controller` parameter to the to the `deployment.yaml` file that deploys the operator.
- You have admin rights to install the Custom Resource Definition (CRD) for Operator.

## Operator Install

Operator has two distinct modes:

> - **Basic**: Operator in basic mode installs Spinnaker into a single namespace without `ValidatingAdmissionWebhook` for doing preflight checks.

Reviewer: ```suggestion
- **Basic**: for installing the Operator and Spinnaker in a single namespace without doing pre-flight checks.
```

Reviewer: I don't think anyone reading this will care if it's a validating admissions webhook or not, just that we won't do pre-flight checks before applying the manifest. Thoughts?

> - **Cluster**: Operator in cluster mode installs Spinnaker across namespaces with `ValidatingAdmissionWebhook` for doing preflight checks. This mode requires a `ClusterRole`.

Reviewer: ```suggestion
- **Cluster**: which allows Operator to install Spinnaker across namespaces and provides pre-flight checks to prevent common misconfigurations. This mode requires a `ClusterRole`.
```

spinnaker-operator-7cd659654b-4vktl   2/2           Running      0             6s
```

> If you want to use a namespace other than `spinnaker-operator` in cluster mode, you'll also need to edit the namespace in `deploy/operator/cluster/role_binding.yaml`.

## Spinnaker Install

To use the examples, change the parameters you need (especially the `persistentStorage` section). To install a basic version of Spinnaker in the "spinnaker" namespace, run the following command:

```bash
kubectl create ns spinnaker
kubectl -n spinnaker apply -f deploy/spinnaker/basic

# Watch the install progress and check out the pods being created too!
kubectl -n spinnaker get spinsvc spinnaker -w
```

### How it works
Spinnaker's configuration can be found in a `spinnakerservices.spinnaker.armory.io` Custom Resource Definition (CRD) that can be stored in version control. After you install Spinnaker Operator, you can use `kubectl` to manage the lifecycle of your deployment.

```yaml
apiVersion: spinnaker.armory.io/v1alpha2
kind: SpinnakerService
metadata:
  name: spinnaker
spec:
  spinnakerConfig:
    config:
      version: 2.21.0
```

> See [the full format](../operator-reference)

Reviewer: Feels like we should link to the patch files repo so that users can see some common setups, in addition to the raw operator reference:
https://github.com/armory/spinnaker-kustomize-patches

kubectl -n <namespace> delete spinnakerservice spinnaker
```

## Managing Configuration

### Kustomize
Because Spinnaker's configuration is now a Kubernetes manifests, we can manage `SpinnakerService` and related manifests in a consistent and repeatable way with [kustomize](https://kustomize.io/).

An example can be found [here](https://github.com/armory-io/spinnaker-operator/tree/master/deploy/spinnaker/kustomize).

```bash
kubectl create ns spinnaker
kustomize build deploy/spinnaker/kustomize | kubectl -n spinnaker apply -f -
```

There are many more possibilities such as:
- managing manifests of MySQL instances
- ensuring the same configuration is used between Staging and Production Spinnaker

> - splitting accounts in their own kustomization for an easy to maintain configuration 

Reviewer: This would also be a good section to call out the spinnaker-patch-files repo:
https://github.com/armory/spinnaker-kustomize-patches

### Secret Management

> You can store secrets in one of the [support secret engine](../../spinnaker-install-admin-guides/secrets/secrets/#supported-secret-engines). 

Reviewer: ```suggestion
You can store secrets in one of the [supported secrets engines](../../spinnaker-install-admin-guides/secrets/secrets/#supported-secret-engines). 
```

> #### Kubernetes Secret

Reviewer: ```suggestion
#### Kubernetes Secrets
```

With the operator, you can also reference secrets stored in existing Kubernetes secrets in the same namespace as Spinnaker.

The format is:

> - `encrypted:k8s!n:<secret name>!k:<secret key>` for string values. These will be added as environment variable to the Spinnaker deployment.

Reviewer: ```suggestion
- `encrypted:k8s!n:<secret name>!k:<secret key>` for string values. This will be added as an environment variable to the Spinnaker deployment.
```

- `encryptedFile:k8s!n:<secret name>!k:<secret key>` for file references. Files will come from a volume mount in the Spinnaker deployment.

#### Custom Halyard Configuration

To override Halyard's configuration, create a `ConfigMap` with the configuration changes you need. For example, if using [secrets management with Vault]({{< ref "secrets-vault" >}}), Halyard and Operator containers will need your Vault configuration:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: halyard-custom-config
data:
  halyard-local.yml: |
    secrets:
      vault:
        enabled: true
        url: <URL of vault server>
        path: <cluster path>
        role: <k8s role>
        authMethod: KUBERNETES
```

You can then mount it in the operator deployment and make it available to the Halyard and Operator containers:

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: spinnaker-operator
  ...
spec:
  template:
    spec:
      containers:
      - name: spinnaker-operator
        ...
        volumeMounts:
        - mountPath: /opt/spinnaker/config/halyard.yml
          name: halconfig-volume
          subPath: halyard-local.yml

>       - name: halyard

Reviewer 2: Curious, Halyard might need this to be set, i'm not 100% sure tho. I know we have it set in our [prod operator config](https://github.com/armory-io/armory-infrastructure/blob/master/spinnaker/prod-operator/deployment.yaml#L40-L42)

```yaml
          env:
          - name: SPRING_PROFILE_ACTIVE
            value: local
```

Author: I don't believe so (I think local is standard but I'm not sure). This part is a carryover from the existing doc and I'll let others continue to improve and correct the doc once it's reorganized.
